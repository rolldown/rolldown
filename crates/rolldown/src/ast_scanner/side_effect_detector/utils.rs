use oxc::{
  ast::{
    ast::{Expression, MemberExpression},
    comments_range, Comment, CommentKind,
  },
  semantic::ReferenceId,
  span::{Atom, Span},
  syntax::operator::{BinaryOperator, LogicalOperator, UnaryOperator, UpdateOperator},
};
use rolldown_common::AstScopes;
use rolldown_ecmascript::ExpressionExt;

use super::SideEffectDetector;

impl<'a> SideEffectDetector<'a> {
  /// Get the nearest comment before the `span`, return `None` if no leading comment is founded.
  ///
  ///  # Examples
  /// ```javascript
  /// /* valid comment for `a`  */ let a = 1;
  ///
  /// // valid comment for `b`
  /// let b = 1;
  ///
  /// // valid comment for `c`
  ///
  ///
  /// let c = 1;
  ///
  /// let d = 1; /* valid comment for `e` */
  /// let e = 2
  /// ```
  /// Derived from https://github.com/oxc-project/oxc/blob/147864cfeb112df526bb83d5b8671b465c005066/crates/oxc_linter/src/utils/tree_shaking.rs#L204
  pub fn leading_comment_for(&self, span: Span) -> Option<(&Comment, &str)> {
    let comment = comments_range(self.comments, ..span.start).next_back()?;

    let comment_text = comment.span.source_text(self.source);
    // If there are non-whitespace characters between the `comment` and the `span`,
    // we treat the `comment` not belongs to the `span`.
    let range_text = Span::new(comment.span.end, span.start).source_text(self.source);
    let only_whitespace = match comment.kind {
      CommentKind::Line => range_text.trim().is_empty(),
      CommentKind::Block => {
        range_text
          .strip_prefix("*/") // for multi-line comment
          .is_some_and(|s| s.trim().is_empty())
      }
    };
    if !only_whitespace {
      return None;
    }

    Some((comment, comment_text))
  }
}

#[derive(PartialEq, Eq, Copy, Clone)]
pub(crate) enum PrimitiveType {
  Null,
  Undefined,
  Boolean,
  Number,
  String,
  BigInt,
  Mixed,
  Unknown,
}

fn merged_known_primitive_types(
  scope: &AstScopes,
  left: &Expression,
  right: &Expression,
) -> PrimitiveType {
  let left_type = known_primitive_type(scope, left);
  if left_type == PrimitiveType::Unknown {
    return PrimitiveType::Unknown;
  }
  let right_type = known_primitive_type(scope, right);
  if right_type == PrimitiveType::Unknown {
    return PrimitiveType::Unknown;
  }
  if right_type == left_type {
    return right_type;
  }
  PrimitiveType::Mixed
}

#[allow(clippy::too_many_lines)]
pub(crate) fn known_primitive_type(scope: &AstScopes, expr: &Expression) -> PrimitiveType {
  match expr {
    Expression::NullLiteral(_) => PrimitiveType::Null,
    Expression::Identifier(id)
      if id.name == "undefined" && scope.is_unresolved(id.reference_id.get().unwrap()) =>
    {
      PrimitiveType::Undefined
    }
    Expression::BooleanLiteral(_) => PrimitiveType::Boolean,
    Expression::NumericLiteral(_) => PrimitiveType::Number,
    Expression::StringLiteral(_) => PrimitiveType::String,
    Expression::BigIntLiteral(_) => PrimitiveType::BigInt,
    Expression::TemplateLiteral(e) => {
      if e.expressions.is_empty() {
        PrimitiveType::String
      } else {
        PrimitiveType::Unknown
      }
    }
    Expression::UpdateExpression(e) => {
      match e.operator {
        UpdateOperator::Increment | UpdateOperator::Decrement => {
          PrimitiveType::Mixed // Can be number or bigint
        }
      }
    }
    Expression::UnaryExpression(e) => match e.operator {
      UnaryOperator::Void => PrimitiveType::Undefined,
      UnaryOperator::Typeof => PrimitiveType::String,
      UnaryOperator::LogicalNot | UnaryOperator::Delete => PrimitiveType::Boolean,
      UnaryOperator::UnaryPlus => PrimitiveType::Number, // Cannot be bigint because that throws an exception
      UnaryOperator::UnaryNegation | UnaryOperator::BitwiseNot => {
        let value = known_primitive_type(scope, &e.argument);
        if value == PrimitiveType::BigInt {
          return PrimitiveType::BigInt;
        }
        if value != PrimitiveType::Unknown && value != PrimitiveType::Mixed {
          return PrimitiveType::Number;
        }
        PrimitiveType::Mixed // Can be number or bigint
      }
    },
    Expression::LogicalExpression(e) => match e.operator {
      LogicalOperator::Or | LogicalOperator::And => {
        merged_known_primitive_types(scope, &e.left, &e.right)
      }
      LogicalOperator::Coalesce => {
        let left = known_primitive_type(scope, &e.left);
        let right = known_primitive_type(scope, &e.right);
        if left == PrimitiveType::Null || left == PrimitiveType::Undefined {
          return right;
        }
        if left != PrimitiveType::Unknown {
          if left != PrimitiveType::Mixed {
            return left; // Definitely not null or undefined
          }
          if right != PrimitiveType::Unknown {
            return PrimitiveType::Mixed; // Definitely some kind of primitive
          }
        }
        PrimitiveType::Unknown
      }
    },
    Expression::BinaryExpression(e) => match e.operator {
      BinaryOperator::StrictEquality
      | BinaryOperator::StrictInequality
      | BinaryOperator::Equality
      | BinaryOperator::Inequality
      | BinaryOperator::LessThan
      | BinaryOperator::GreaterThan
      | BinaryOperator::GreaterEqualThan
      | BinaryOperator::LessEqualThan
      | BinaryOperator::Instanceof
      | BinaryOperator::In => PrimitiveType::Boolean,
      BinaryOperator::Addition => {
        let left = known_primitive_type(scope, &e.left);
        let right = known_primitive_type(scope, &e.right);
        if left == PrimitiveType::String || right == PrimitiveType::String {
          PrimitiveType::String
        } else if left == PrimitiveType::BigInt && right == PrimitiveType::BigInt {
          PrimitiveType::BigInt
        } else if !matches!(
          left,
          PrimitiveType::Unknown | PrimitiveType::Mixed | PrimitiveType::BigInt
        ) && !matches!(
          right,
          PrimitiveType::Unknown | PrimitiveType::Mixed | PrimitiveType::BigInt
        ) {
          PrimitiveType::Number
        } else {
          PrimitiveType::Mixed // Can be number or bigint or string (or an exception)
        }
      }
      BinaryOperator::Subtraction
      | BinaryOperator::Multiplication
      | BinaryOperator::Division
      | BinaryOperator::Remainder
      | BinaryOperator::Exponential
      | BinaryOperator::BitwiseAnd
      | BinaryOperator::BitwiseOR
      | BinaryOperator::ShiftRight
      | BinaryOperator::ShiftLeft
      | BinaryOperator::ShiftRightZeroFill
      | BinaryOperator::BitwiseXOR => PrimitiveType::Mixed,
    },

    Expression::AssignmentExpression(e) => match e.operator {
      oxc::syntax::operator::AssignmentOperator::Assign => known_primitive_type(scope, &e.right),
      oxc::syntax::operator::AssignmentOperator::Addition => {
        let right = known_primitive_type(scope, &e.right);
        if right == PrimitiveType::String {
          PrimitiveType::String
        } else {
          PrimitiveType::Mixed // Can be number or bigint or string (or an exception)
        }
      }
      oxc::syntax::operator::AssignmentOperator::Subtraction
      | oxc::syntax::operator::AssignmentOperator::Multiplication
      | oxc::syntax::operator::AssignmentOperator::Division
      | oxc::syntax::operator::AssignmentOperator::Remainder
      | oxc::syntax::operator::AssignmentOperator::ShiftLeft
      | oxc::syntax::operator::AssignmentOperator::ShiftRight
      | oxc::syntax::operator::AssignmentOperator::ShiftRightZeroFill
      | oxc::syntax::operator::AssignmentOperator::BitwiseOR
      | oxc::syntax::operator::AssignmentOperator::BitwiseXOR
      | oxc::syntax::operator::AssignmentOperator::BitwiseAnd
      | oxc::syntax::operator::AssignmentOperator::LogicalAnd
      | oxc::syntax::operator::AssignmentOperator::LogicalOr
      | oxc::syntax::operator::AssignmentOperator::LogicalNullish
      | oxc::syntax::operator::AssignmentOperator::Exponential => PrimitiveType::Mixed,
    },
    _ => PrimitiveType::Unknown,
  }
}

pub fn can_change_strict_to_loose(scope: &AstScopes, a: &Expression, b: &Expression) -> bool {
  let x = known_primitive_type(scope, a);
  let y = known_primitive_type(scope, b);
  x == y && !matches!(x, PrimitiveType::Unknown | PrimitiveType::Mixed)
}

pub fn is_primitive_literal(scope: &AstScopes, expr: &Expression) -> bool {
  match expr {
    Expression::NullLiteral(_)
    | Expression::BooleanLiteral(_)
    | Expression::NumericLiteral(_)
    | Expression::StringLiteral(_)
    | Expression::BigIntLiteral(_) => true,
    // Include `+1` / `-1`.
    Expression::UnaryExpression(e)
      if matches!(e.operator, |UnaryOperator::UnaryNegation| UnaryOperator::UnaryPlus)
        && matches!(e.argument, Expression::NumericLiteral(_)) =>
    {
      true
    }
    Expression::Identifier(id)
      if id.name == "undefined" && scope.is_unresolved(id.reference_id.get().unwrap()) =>
    {
      true
    }
    _ => false,
  }
}

pub fn extract_member_expr_chain<'a>(
  expr: &'a MemberExpression,
  max_len: usize,
) -> Option<(ReferenceId, Vec<Atom<'a>>)> {
  if max_len == 0 {
    return None;
  }
  let mut chain = vec![];
  match expr {
    MemberExpression::ComputedMemberExpression(computed_expr) => {
      let Expression::StringLiteral(ref str) = computed_expr.expression else {
        return None;
      };
      chain.push(str.value.clone());
      let mut cur = &computed_expr.object;
      extract_rest_member_expr_chain(&mut cur, &mut chain, max_len).map(|ref_id| (ref_id, chain))
    }
    MemberExpression::StaticMemberExpression(static_expr) => {
      let mut cur = &static_expr.object;
      chain.push(static_expr.property.name.clone());
      extract_rest_member_expr_chain(&mut cur, &mut chain, max_len).map(|ref_id| (ref_id, chain))
    }
    MemberExpression::PrivateFieldExpression(_) => None,
  }
}

fn extract_rest_member_expr_chain<'a>(
  cur: &mut &'a Expression,
  chain: &mut Vec<Atom<'a>>,
  max_len: usize,
) -> Option<ReferenceId> {
  loop {
    match &cur {
      Expression::StaticMemberExpression(expr) => {
        *cur = &expr.object;
        chain.push(expr.property.name.clone());
      }
      Expression::ComputedMemberExpression(expr) => {
        let Expression::StringLiteral(ref str) = expr.expression else {
          break;
        };
        chain.push(str.value.clone());
        *cur = &expr.object;
      }
      Expression::Identifier(ident) => {
        chain.push(ident.name.clone());
        let ref_id = ident.reference_id.get().expect("should have reference_id");
        chain.reverse();
        return Some(ref_id);
      }
      _ => break,
    }
    // If chain exceeds the max length, that means we are not interest in this member expression.
    // return `None`
    if chain.len() >= max_len {
      return None;
    }
  }
  None
}

/// https://github.com/evanw/esbuild/blob/d34e79e2a998c21bb71d57b92b0017ca11756912/internal/js_ast/js_ast_helpers.go#L2594-L2639
pub fn is_side_effect_free_unbound_identifier_ref(
  scope: &AstScopes,
  value: &Expression,
  guard_condition: &Expression,
  mut is_yes_branch: bool,
) -> Option<bool> {
  let ident = value.as_identifier()?;
  let is_unresolved = scope.is_unresolved(ident.reference_id()?);
  if !is_unresolved {
    return Some(false);
  }
  let bin_expr = guard_condition.as_binary_expression()?;
  match bin_expr.operator {
    BinaryOperator::StrictEquality
    | BinaryOperator::StrictInequality
    | BinaryOperator::Equality
    | BinaryOperator::Inequality => {
      let (mut ty_of, mut string) = (&bin_expr.left, &bin_expr.right);
      if matches!(ty_of, Expression::StringLiteral(_)) {
        std::mem::swap(&mut string, &mut ty_of);
      }
      let unary = ty_of.as_unary_expression()?;
      if !(unary.operator == UnaryOperator::Typeof
        && matches!(unary.argument, Expression::Identifier(_)))
      {
        return Some(false);
      }
      let string = string.as_string_literal()?;

      if (string.value.eq("undefined") == is_yes_branch)
        == matches!(
          bin_expr.operator,
          BinaryOperator::Inequality | BinaryOperator::StrictInequality
        )
      {
        let type_of_value = unary.argument.as_identifier()?;
        if type_of_value.name == ident.name {
          return Some(true);
        }
      }
    }
    BinaryOperator::LessThan
    | BinaryOperator::LessEqualThan
    | BinaryOperator::GreaterThan
    | BinaryOperator::GreaterEqualThan => {
      let (mut ty_of, mut string) = (&bin_expr.left, &bin_expr.right);
      if matches!(ty_of, Expression::StringLiteral(_)) {
        std::mem::swap(&mut string, &mut ty_of);
        is_yes_branch = !is_yes_branch;
      }

      let unary = ty_of.as_unary_expression()?;
      if !(unary.operator == UnaryOperator::Typeof
        && matches!(unary.argument, Expression::Identifier(_)))
      {
        return Some(false);
      }

      let string = string.as_string_literal()?;

      if string.value == "u"
        && is_yes_branch
          == matches!(bin_expr.operator, BinaryOperator::LessThan | BinaryOperator::LessEqualThan)
      {
        let type_of_value = unary.argument.as_identifier()?;
        if type_of_value.name == ident.name {
          return Some(true);
        }
      }
    }
    _ => {}
  }
  Some(false)
}
