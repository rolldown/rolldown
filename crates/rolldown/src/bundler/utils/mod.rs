pub mod ast_scope;
pub mod bitset;
pub mod renamer;
pub mod resolve_id;
