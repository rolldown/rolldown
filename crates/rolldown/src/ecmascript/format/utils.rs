use crate::types::generator::GenerateContext;
use arcstr::ArcStr;
use rolldown_common::OutputExports;
use rolldown_error::{BuildDiagnostic, DiagnosableResult};
use rolldown_utils::ecma_script::is_validate_assignee_identifier_name;

/// According to the amount of `.` in the name (levels),
/// it generates the initialization code and the final code.
///
/// # Example
///
/// for a IIFE named `namespace.module.hello`, it will generate:
///
/// - The initialization code:
///    ```js
///    this.namespace = this.namespace || {};
///    this.namespace.module = this.namespace.module || {};
///    ```
///  - The final code:
///    ```js
///    this.namespace.module.hello
///    ```
fn generate_namespace_definition(name: &str) -> (String, String) {
  let parts: Vec<&str> = name.split('.').collect();

  let initialization_code = parts
    .iter()
    .enumerate()
    .scan(String::new(), |state, (i, part)| {
      // We use `scan` to generate the declaration sentence level-by-level.
      let callee = generate_callee(part);
      state.push_str(&callee);
      let line = if i < parts.len() - 1 {
        Some(format!("this{state} = this{state} || {{}};\n"))
      } else {
        None
      };
      Some(line)
    })
    .flatten()
    .collect::<String>();

  // TODO do not call the `generate_callee` function twice.
  let final_code =
    format!("this{}", parts.iter().map(|&part| generate_callee(part)).collect::<String>());

  (initialization_code, final_code)
}

/// This function generates a namespace definition for the given name, especially for IIFE format or UMD format.
/// If the name contains a dot, it will be regarded as a namespace definition.
/// Otherwise, it will be regarded as a variable definition.
pub fn generate_identifier(ctx: &mut GenerateContext<'_>) -> DiagnosableResult<(String, String)> {
  if let Some(name) = &ctx.options.name {
    // It is same as Rollup.
    if name.contains('.') {
      let (decl, expr) = generate_namespace_definition(name);
      Ok((
        decl,
        if ctx.options.extend && matches!(&ctx.options.exports, OutputExports::Named) {
          format!("{expr} = {expr} || {{}}")
        } else {
          expr
        },
      ))
    } else if ctx.options.extend {
      if matches!(ctx.options.exports, OutputExports::Named) {
        Ok((String::new(), format!("this{name} = this{name} || {{}}")))
      } else {
        Ok((String::new(), format!("this{name}")))
      }
    } else if is_validate_assignee_identifier_name(name) {
      Ok((String::new(), format!("var {name}")))
    } else {
      // This behavior is aligned with Rollup. If using `output.extend: true`, this error won't be triggered.
      let name = ArcStr::from(name);
      Err(vec![BuildDiagnostic::illegal_identifier_as_name(name)])
    }
  } else {
    // If the `name` is empty, you may be impossible to call the result.
    // But it is normal if we do not have exports.
    // However, if there is no export, it is recommended to use `app` format.
    ctx
      .warnings
      .push(BuildDiagnostic::missing_name_option_for_iife_export().with_severity_warning());
    Ok((String::new(), String::new()))
  }
}

// It is a helper function to generate a callee for the given name.
fn generate_callee(name: &str) -> String {
  if is_validate_assignee_identifier_name(name) {
    format!(".{name}")
  } else {
    format!("[\"{name}\"]")
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn test_generate_namespace_definition() {
    let result = generate_namespace_definition("a.b.c");
    assert_eq!(result.0, "this.a = this.a || {};\nthis.a.b = this.a.b || {};\n");
    assert_eq!(result.1, "this.a.b.c");
  }

  #[test]
  fn test_reserved_identifier_as_name() {
    let result = generate_namespace_definition("1.2.3");
    assert_eq!(
      result.0,
      "this[\"1\"] = this[\"1\"] || {};\nthis[\"1\"][\"2\"] = this[\"1\"][\"2\"] || {};\n"
    );
    assert_eq!(result.1, "this[\"1\"][\"2\"][\"3\"]");
  }

  #[test]
  /// It is related a bug in rollup. Check it out in [rollup/rollup#5603](https://github.com/rollup/rollup/issues/5603).
  fn test_invalid_identifier_as_name() {
    let result = generate_namespace_definition("toString.valueOf.constructor");
    assert_eq!(result.0, "this.toString = this.toString || {};\nthis.toString.valueOf = this.toString.valueOf || {};\n");
    assert_eq!(result.1, "this.toString.valueOf.constructor");
  }
}
