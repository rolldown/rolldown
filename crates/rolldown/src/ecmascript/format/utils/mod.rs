use itertools::Itertools;
use rolldown_std_utils::OptionExt;

use crate::{
  types::generator::GenerateContext,
  utils::chunk::collect_render_chunk_imports::{
    collect_render_chunk_imports, ExternalRenderImportStmt, RenderImportDeclarationSpecifier,
    RenderImportStmt,
  },
};

pub mod namespace;

pub fn render_factory_parameters(
  ctx: &mut GenerateContext<'_>,
  externals: &[ExternalRenderImportStmt],
  has_exports: bool,
) -> String {
  let mut parameters = if has_exports { vec!["exports"] } else { vec![] };
  externals.iter().for_each(|external| {
    let symbol_name = ctx.chunk.canonical_name_by_token.get(&external.binding_name_token).unpack();
    parameters.push(symbol_name.as_str());
  });
  parameters.join(", ")
}

pub fn render_chunk_external_imports(
  ctx: &GenerateContext<'_>,
) -> (String, Vec<ExternalRenderImportStmt>) {
  let render_import_stmts =
    collect_render_chunk_imports(ctx.chunk, ctx.link_output, ctx.chunk_graph);

  let mut import_code = String::new();
  let externals = render_import_stmts
    .into_iter()
    .filter_map(|stmt| {
      if let RenderImportStmt::ExternalRenderImportStmt(external_stmt) = stmt {
        let symbol_name =
          ctx.chunk.canonical_name_by_token.get(&external_stmt.binding_name_token).unpack();
        match &external_stmt.specifiers {
          RenderImportDeclarationSpecifier::ImportSpecifier(specifiers) => {
            // Empty specifiers can be ignored in IIFE.
            if specifiers.is_empty() {
              None
            } else {
              let specifiers = specifiers
                .iter()
                .map(|specifier| {
                  if let Some(alias) = &specifier.alias {
                    format!("{}: {alias}", specifier.imported)
                  } else {
                    specifier.imported.to_string()
                  }
                })
                .collect::<Vec<_>>();
              import_code
                .push_str(&format!("const {{ {} }} = {symbol_name};\n", specifiers.join(", ")));
              Some(external_stmt)
            }
          }
          RenderImportDeclarationSpecifier::ImportStarSpecifier(alias) => {
            import_code.push_str(&format!("const {alias} = {symbol_name};\n"));
            Some(external_stmt)
          }
        }
      } else {
        None
      }
    })
    .collect_vec();

  (import_code, externals)
}
