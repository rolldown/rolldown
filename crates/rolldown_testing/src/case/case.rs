use std::{borrow::Cow, path::Path};

use crate::utils::RUNTIME_MODULE_OUTPUT_RE;

use super::fixture::Fixture;
use rolldown::BundleOutput;
use rolldown_common::Output;
use rolldown_error::{BuildDiagnostic, DiagnosticOptions};
use rolldown_sourcemap::SourcemapVisualizer;

pub struct Case {
  fixture: Fixture,
  snapshot: String,
}

impl Case {
  pub fn new(path: impl AsRef<Path>) -> Self {
    // Paths could be UNC format in windows, see https://github.com/rust-lang/rust/issues/42869 for more details
    let path = dunce::simplified(path.as_ref());

    Self { fixture: Fixture::new(path.to_path_buf()), snapshot: String::new() }
  }

  pub fn run(self) {
    std::env::set_var("ROLLDOWN_TEST", "1");
    tokio::runtime::Runtime::new().unwrap().block_on(self.run_inner());
  }

  pub async fn run_inner(mut self) {
    let build_output = self.fixture.bundle(true, false).await;
    if build_output.errors.is_empty() {
      assert!(!self.fixture.test_config().expect_error, "expected error, but got success");
      self.render_assets_to_snapshot(build_output);
    } else {
      assert!(
        self.fixture.test_config().expect_error,
        "expected success, but got errors: {:?}",
        build_output.errors
      );
      self.render_errors_to_snapshot(build_output.errors);
    }
    self.make_snapshot();
    self.fixture.exec();
  }

  fn render_assets_to_snapshot(&mut self, outputs: BundleOutput) {
    let mut assets = outputs.assets;
    // Make the snapshot consistent
    let mut warnings = outputs.warnings;
    warnings.sort_by(|a, b| {
      let a = a.to_string();
      let b = b.to_string();
      a.cmp(&b)
    });
    if !warnings.is_empty() {
      self.snapshot.push_str("# warnings\n\n");
      let diagnostics = warnings.into_iter().map(|e| {
        (
          e.kind(),
          e.into_diagnostic_with(&DiagnosticOptions { cwd: self.fixture.dir_path().to_path_buf() }),
        )
      });
      let rendered = diagnostics
        .flat_map(|(code, diagnostic)| {
          [
            Cow::Owned(format!("## {code}\n")),
            "```text".into(),
            Cow::Owned(diagnostic.to_string()),
            "```".into(),
          ]
        })
        .collect::<Vec<_>>()
        .join("\n");
      self.snapshot.push_str(&rendered);
      self.snapshot.push('\n');
    }

    self.snapshot.push_str("# Assets\n\n");
    assets.sort_by_key(|c| c.filename().to_string());
    let artifacts = assets
      .iter()
      .filter(|asset| !asset.filename().contains("$runtime$") && matches!(asset, Output::Chunk(_)))
      .flat_map(|asset| {
        let content = std::str::from_utf8(asset.content_as_bytes()).unwrap();
        let content = if self.fixture.test_config().hidden_runtime_module {
          RUNTIME_MODULE_OUTPUT_RE.replace_all(content, "")
        } else {
          Cow::Borrowed(content)
        };

        [Cow::Owned(format!("## {}\n", asset.filename())), "```js".into(), content, "```".into()]
      })
      .collect::<Vec<_>>()
      .join("\n");
    self.snapshot.push_str(&artifacts);

    if self.fixture.test_config().snapshot_output_stats {
      self.render_stats_to_snapshot(&assets);
    }

    if self.fixture.test_config().visualize_sourcemap {
      self.render_sourcemap_visualizer_to_snapshot(&assets);
    }
  }

  fn render_stats_to_snapshot(&mut self, assets: &[Output]) {
    self.snapshot.push_str("\n\n## Output Stats\n\n");
    let stats = assets
      .iter()
      .flat_map(|asset| match asset {
        Output::Chunk(chunk) => {
          vec![Cow::Owned(format!(
            "- {}, is_entry {}, is_dynamic_entry {}, exports {:?}",
            chunk.filename.as_str(),
            chunk.is_entry,
            chunk.is_dynamic_entry,
            chunk.exports
          ))]
        }
        Output::Asset(_) => vec![],
      })
      .collect::<Vec<_>>()
      .join("\n");
    self.snapshot.push_str(&stats);
  }

  fn render_errors_to_snapshot(&mut self, mut errors: Vec<BuildDiagnostic>) {
    self.snapshot.push_str("# Errors\n\n");
    errors.sort_by_key(|e| e.kind().to_string());
    let diagnostics = errors.into_iter().map(|e| {
      (
        e.kind(),
        e.into_diagnostic_with(&DiagnosticOptions { cwd: self.fixture.dir_path().to_path_buf() }),
      )
    });

    let rendered = diagnostics
      .flat_map(|(code, diagnostic)| {
        [
          Cow::Owned(format!("## {code}\n")),
          "```text".into(),
          Cow::Owned(diagnostic.to_string()),
          "```".into(),
        ]
      })
      .collect::<Vec<_>>()
      .join("\n");
    self.snapshot.push_str(&rendered);
  }

  fn render_sourcemap_visualizer_to_snapshot(&mut self, assets: &[Output]) {
    self.snapshot.push_str("\n\n# Sourcemap Visualizer\n\n");
    let visualizer_result = assets
      .iter()
      .filter_map(|asset| match asset {
        Output::Chunk(chunk) => chunk
          .map
          .as_ref()
          .map(|sourcemap| SourcemapVisualizer::new(&chunk.code, sourcemap).into_visualizer_text()),
        Output::Asset(_) => None,
      })
      .collect::<Vec<_>>()
      .join("\n");
    self.snapshot.push_str(&visualizer_result);
  }

  fn make_snapshot(&self) {
    // Configure insta to use the fixture path as the snapshot path
    let fixture_folder = self.fixture.dir_path();
    let mut settings = insta::Settings::clone_current();
    let content = self.snapshot.to_string();
    settings.set_snapshot_path(fixture_folder);
    settings.set_prepend_module_to_snapshot(false);
    settings.set_input_file(fixture_folder);
    settings.bind(|| {
      insta::assert_snapshot!("artifacts", content);
    });
  }
}
