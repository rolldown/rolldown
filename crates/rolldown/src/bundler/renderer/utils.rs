use std::borrow::Cow;

use oxc::span::{Atom, Span};
use rolldown_common::SymbolRef;
use string_wizard::UpdateOptions;

use crate::bundler::{
  linker::linker_info::LinkingInfo,
  module::{Module, NormalModule},
};

use super::{AstRenderContext, AstRenderer};
impl<'r> AstRenderContext<'r> {
  pub fn canonical_name_for(&self, symbol: SymbolRef) -> &'r Atom {
    self.graph.symbols.canonical_name_for(symbol, self.canonical_names)
  }

  pub fn need_to_rename(&self, symbol: SymbolRef) -> Option<&Atom> {
    let canonical_ref = self.graph.symbols.par_canonical_ref_for(symbol);
    self.canonical_names.get(&canonical_ref)
  }

  pub fn canonical_name_for_runtime(&self, name: &str) -> &Atom {
    let symbol = self.graph.runtime.resolve_symbol(name);
    self.canonical_name_for(symbol)
  }

  pub fn importee_by_span(&self, span: Span) -> &Module {
    &self.graph.modules[self.module.importee_id_by_span(span)]
  }

  pub fn remove_node(&mut self, span: Span) {
    self.source.remove(span.start, span.end);
  }

  pub fn hoisted_module_declaration(&mut self, decl_start: u32, content: String) {
    let start = self.first_stmt_start.unwrap_or(decl_start);
    self.source.append_left(start, content);
  }

  pub fn generate_import_commonjs_module(
    &self,
    importee: &NormalModule,
    importee_linking_info: &LinkingInfo,
    with_declaration: bool,
  ) -> String {
    let wrap_ref_name = self.canonical_name_for(importee_linking_info.wrapper_ref.unwrap());
    let to_esm_ref_name = self.canonical_name_for_runtime("__toESM");
    let code = format!(
      "{to_esm_ref_name}({wrap_ref_name}(){})",
      if self.module.module_type.is_esm() { ", 1" } else { "" }
    );
    if with_declaration {
      let symbol_ref =
        self.linking_info.local_symbol_for_import_cjs.get(&importee.id).copied().unwrap_or_else(
          || {
            panic!(
              "Cannot find local symbol for importee: {:?} with importer {:?} {:?}",
              importee.resource_id, self.module.resource_id, self.module.exports_kind
            )
          },
        );
      let final_name = self.canonical_name_for(symbol_ref);
      format!("var {final_name} = {code};\n")
    } else {
      code
    }
  }
}

impl<'r> AstRenderer<'r> {
  pub fn is_global_require(&self, callee: &oxc::ast::ast::Expression<'_>) -> bool {
    matches!(callee, oxc::ast::ast::Expression::Identifier(ident) if
      ident.name == "require"
      && self.ctx.module.scope.is_unresolved(ident.reference_id.get().unwrap())
    )
  }

  pub fn overwrite(&mut self, start: u32, end: u32, content: String) {
    self.ctx.source.update_with(
      start,
      end,
      content,
      UpdateOptions { overwrite: true, ..Default::default() },
    );
  }

  pub fn generate_namespace_variable_declaration(&mut self) -> Option<String> {
    if self.ctx.module.is_namespace_referenced() {
      let ns_name = self.ctx.canonical_name_for(self.ctx.module.namespace_symbol);
      let exports: String = self
        .ctx
        .linking_info
        .sorted_exports()
        .map(|(exported_name, resolved_export)| {
          let canonical_ref =
            self.ctx.graph.symbols.par_canonical_ref_for(resolved_export.symbol_ref);
          let symbol = self.ctx.graph.symbols.get(canonical_ref);
          let return_expr = if let Some(ns_alias) = &symbol.namespace_alias {
            let canonical_ns_name = &self.ctx.canonical_names[&ns_alias.namespace_ref];
            format!("{canonical_ns_name}.{exported_name}",)
          } else {
            let canonical_name = self.ctx.canonical_names.get(&canonical_ref).unwrap();
            format!("{canonical_name}",)
          };
          format!("{}get {exported_name}() {{ return {return_expr} }}", self.indentor)
        })
        .collect::<Vec<_>>()
        .join(",\n");
      Some(format!("var {ns_name} = {{\n{exports}\n}};\n",))
    } else {
      None
    }
  }

  pub fn remove_node(&mut self, span: Span) {
    self.ctx.source.remove(span.start, span.end);
  }

  pub fn rewrite_symbol(
    &mut self,
    symbol_ref: SymbolRef,
    original_name: &Atom,
    pos: Span,
    is_callee: bool,
  ) {
    let canonical_ref = self.ctx.graph.symbols.par_canonical_ref_for(symbol_ref);
    let symbol = self.ctx.graph.symbols.get(canonical_ref);
    let rendered_symbol = if let Some(ns_alias) = &symbol.namespace_alias {
      // If import symbol from commonjs, the reference of the symbol is not resolved,
      // Here need write it to property access. eg `import { a } from 'cjs'; console.log(a)` => `console.log(import_a.a)`
      // Note: we should rewrite call expression to indirect call, eg `import { a } from 'cjs'; console.log(a())` => `console.log((0, import_a.a)())`
      let canonical_ns_name = self.canonical_name_for(ns_alias.namespace_ref);
      let content = format!("{canonical_ns_name}.{}", ns_alias.property_name);
      Cow::Owned(content)
    } else if let Some(name) = self.need_to_rename(symbol_ref) {
      Cow::Borrowed(name.as_str())
    } else {
      return;
    };

    if original_name != &rendered_symbol {
      self.ctx.source.update(
        pos.start,
        pos.end,
        if is_callee { format!("(0, {rendered_symbol})",) } else { rendered_symbol.into_owned() },
      );
    }
  }

  pub fn canonical_name_for(&self, symbol: SymbolRef) -> &'r Atom {
    self.ctx.graph.symbols.canonical_name_for(symbol, self.ctx.canonical_names)
  }

  pub fn canonical_name_for_runtime(&self, name: &str) -> &Atom {
    let symbol = self.ctx.graph.runtime.resolve_symbol(&Atom::new_inline(name));
    self.canonical_name_for(symbol)
  }

  pub fn need_to_rename(&self, symbol: SymbolRef) -> Option<&Atom> {
    let canonical_ref = self.ctx.graph.symbols.par_canonical_ref_for(symbol);
    self.ctx.canonical_names.get(&canonical_ref)
  }

  pub fn hoisted_module_declaration(&mut self, decl_start: u32, content: String) {
    let start = self.ctx.first_stmt_start.unwrap_or(decl_start);
    self.ctx.source.append_left(start, content);
  }
}
