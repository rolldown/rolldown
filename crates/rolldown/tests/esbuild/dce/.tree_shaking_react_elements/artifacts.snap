---
source: crates/rolldown_testing/src/case/case.rs
expression: content
input_file: crates/rolldown/tests/esbuild/dce/tree_shaking_react_elements
---
# warnings

## UNRESOLVED_IMPORT

```text
[UNRESOLVED_IMPORT] Warning: "react/jsx-runtime" is imported by "entry.jsx", but could not be resolved â€“ treating it as an external dependency.

```
# Assets

## entry_jsx.mjs

```js
import { Fragment as _Fragment, jsx as _jsx } from "react/jsx-runtime";

//#region entry.jsx
function Foo() {}
let a = _jsx('div', {});
let b = _jsx(Foo, {children: a});
let c = _jsx(_Fragment, {children: b});
let d = _jsx('div', {});
let e = _jsx(Foo, {children: d});
let f = _jsx(_Fragment, {children: e});
console.log(f);

//#endregion
```
